### PR TITLE
react-hooks/rules-of-hooks: Improve support for `do/while` loops

### LINKED ISSUE #31687: Bug: [eslint-plugin-react-hooks] incorrectly reports an error when hook is called outside of a loop.
The following code triggers an ESLint error with the rule `eslintreact-hooks/rules-of-hooks` stating:  
> "React Hook 'useState' may be executed more than once. Possibly because it is called in a loop. React Hooks must be called in the exact same order in every component render."

However, the hook `useState` is not inside the loop, and there is no reason for the error to be thrown.

### React version:
React 19.0.0 (latest stable)
eslint-plugin-react-hooks 5.1.0 (latest stable)

## Steps To Reproduce
1. Create a functional component with `useState` hook.
2. Add a `for` loop inside the component (but outside the hook).
3. Run the ESLint with the `eslint-react-hooks` plugin enabled.

```tsx
const Component = () => {
  const [state, setState] = useState(0);

  for (let i = 0; i < 10; i++) {
    console.log(i);
  }

  return <div></div>;
};
```

### Link to code example:
https://codesandbox.io/p/devbox/8dlj6f
Run `npm run lint`

## The current behavior
ESLint throws an error about `useState` potentially being called multiple times, even though it is not in a loop.

## The expected behavior
No ESLint error should be thrown, as `useState` is not inside the loop, and should follow the rule of hooks correctly.

### PR DESCRIPTION
**Pull request by tyxla**
## Summary

This is an alternative to #28714. 

Reverts the main fix there and implements a different solution by adding special handling for `DoWhileStatement`.

Also adds a unit test that was breaking the previous fix (props to @skratchdot for it). 

Please, refer to the inline comments for more information.

Fixes #31687. Also see https://github.com/facebook/react/pull/28714#issuecomment-2531264046.

## How did you test this change?

I've added unit tests that cover the case and verified that they pass by running:

```
yarn test packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js --watch
```

I've also verified that the rest of the tests continue to pass by running:

```
yarn test
```

and

```
yarn test --prod
```

### CODE REVIEW

File: packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js [hunk 1]
```diff
@@ -550,6 +550,18 @@ const tests = {
       // TODO: this should error but doesn't.
       // errors: [genericError('useState')],
     },
+    {
+      code: normalizeIndent`
+        // Valid because the hook is outside of the loop
+        const Component = () => {
+          const [state, setState] = useState(0);
+          for (let i = 0; i < 10; i++) {
```
Comment (tyxla) [inline packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js:558]: Note that the hook is called in a valid way, outside of the loop.

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 2]
```diff
@@ -100,6 +100,16 @@ function isInsideComponentOrHook(node) {
...
+function isInsideDoWhileLoop(node) {
+  while (node) {
+    if ( node.type === 'DoWhileStatement' ) {
+      return true;
+    }
+    node = node.parent;
+  }
+  return false;
+}
+
```
Comment (tyxla) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js:112]: Adding an extra utility to traverse up the tree to check if the node is called within a `do/while` loop.

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 3]
```diff
@@ -295,7 +305,7 @@ export default {
           if (pathList.has(segment.id)) {
             const pathArray = Array.from(pathList);
             const cyclicSegments = pathArray.slice(
-              pathArray.indexOf(segment.id) - 1,
+              pathArray.indexOf(segment.id) + 1,
```
Comment (tyxla) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js:308]: Reverts the change from #28714 as it was introducing false positives (see https://github.com/facebook/react/issues/31687).

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 4]
```diff
@@ -485,7 +495,7 @@ export default {
           for (const hook of reactHooks) {
             // Report an error if a hook may be called more then once.
             // `use(...)` can be called in loops.
-            if (cycled && !isUseIdentifier(hook)) {
+            if ( (cycled || isInsideDoWhileLoop(hook)) && !isUseIdentifier(hook)) {
```
Comment (tyxla) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js:498]: We're now considering a hook "cycled" also if it's called within a do/while loop.

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 5]
```diff
@@ -520,7 +530,8 @@ export default {
               if (
                 !cycled &&
                 pathsFromStartToEnd !== allPathsFromStartToEnd &&
-                !isUseIdentifier(hook) // `use(...)` can be called conditionally.
+                !isUseIdentifier(hook) && // `use(...)` can be called conditionally.
+                !isInsideDoWhileLoop(hook) // wrapping do/while loops are checked separately.
```
Comment (tyxla) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js:534]: Specifically excluding hooks inside `do/while` calls because we're handling those cases separately in the above  check.

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js (hunk 2 above)
Comment (skratchdot) [reply packages/eslint-plugin-react-hooks/src/RulesOfHooks.js]: yesterday was the first time i looked at the code in this plugin. i'm curious why all the "is hook inside loop" logic isn't handled this way. seems like the same could be done for all types of loops/cycles and invalid code.  i found the "counting segments" code somewhat hard to follow and this function is much more clear to me. maybe the counting segments was done for performance reasons?

without fully understanding the existing code, it seems like this plugin could be rewritten so that we first find all "hook" functions, then call a "isInsideCycle" that climbs up the tree looking for nodes like `DoWhileStatement` or `ForStatement` etc.

anyways, good work on this (and thanks)!

Comment (tyxla) [reply packages/eslint-plugin-react-hooks/src/RulesOfHooks.js]: I agree - it does feel quite complex at first glance, and my initial expectation was that these would be handled top-down, with loops and control structures at the top, and then function/hook calls at the bottom, but I assumed there was a good reason (likely performance) for it to be that way. Anyway, rewriting that would be a major overhaul of the plugin, and one definitely outside of this PR's scope.

**Comment by tyxla** [comment]
@mofeiZ or @josephsavona - could I please ask for a review of this fix? Thanks in advance!

**Review by eps1lon — APPROVED** [review]

**Comment by Mathias-S** [comment]
@eps1lon Can you release a (non-canary) patch with this bug fix?
